### PR TITLE
Run processing algorithms using task manager

### PR DESCRIPTION
**Pull request by nyalldawson**
And make processing algorithm execution safe to run in background threads.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@wonder-sk I'd appreciate a review here :)

Pertinent commits are:

- https://github.com/qgis/QGIS/pull/4799/commits/112a7321a6da3b5151ad80de30fe30940764e645
- https://github.com/qgis/QGIS/pull/4799/commits/d642e5d851d72a37486441c9b21171e656addbbf
- https://github.com/qgis/QGIS/pull/4799/commits/66336fb037d5c207f9b99dca83b40830da535020
- https://github.com/qgis/QGIS/pull/4799/commits/78382c45dfceabab5f0adaa66eef27437a9f1195
- https://github.com/qgis/QGIS/pull/4799/commits/bca03a5778b3bd570eaba86ed89c65846936e50d
- https://github.com/qgis/QGIS/pull/4799/commits/77daede39049a15525768a399fbd58121d2ef465

**Comment by nyalldawson** [comment]
@wonder-sk I believe there's still one thing we need to address here - while we agreed that adding layers to a store with a different thread affinity is a bad thing, I think we still need to make sure that accessing layers from a store (and hence QgsProject) is safe to do from any thread. It'd be a matter of protecting the QMap member in QgsMapLayerStore with some appropriate read/write mutexes so that there's no possible issue if someone tries to retrieve a layer from a different thread while a layer is being added/ removed to the store from the main thread.

**Comment by nirvn** [comment]
OMG, this is the promised land, we're about to reach it! 😉 

This will work so well with your status bar locator feature.

**Comment by nirvn** [comment]
@nyalldawson, testing this PR out -- things are looking good. I however noticed that the progress bar for the zonal stats alg isn't working anymore. The whole UI freezes until the process is over.

File: src/core/processing/qgsprocessingalgorithm.cpp [hunk 1]
```diff
@@ -361,25 +361,74 @@ bool QgsProcessingAlgorithm::runPrepared( QgsProcessingContext &context, QgsProc
   Q_ASSERT_X( mHasPrepared, "QgsProcessingAlgorithm::runPrepared", "prepare() was not called for the algorithm instance" );
   Q_ASSERT_X( !mHasExecuted, "QgsProcessingAlgorithm::runPrepared", "runPrepared() was already called for this algorithm instance" );
 
+  // Hey kids, let's all be thread safe! It's the fun thing to do!
+  //
+  // First, let's see if we're going to run into issues.
+  QgsProcessingContext *runContext = nullptr;
+  if ( context.thread() == QThread::currentThread() )
+  {
+    // OH. No issues. Seems you're running everything in the same thread, so go about your business. Sorry about
```
Comment (luipir) [inline src/core/processing/qgsprocessingalgorithm.cpp]: In this way it's a pleasure read the code :) 
super +1

**Comment by nyalldawson** [comment]
> , testing this PR out -- things are looking good. I however noticed that the progress bar for the zonal stats alg isn't working anymore. The whole UI freezes until the process is over.

That's fixed now

**Comment by gioman** [comment]
@nyalldawson sorry for the naive question, does this applies only to qgis native algs or also any other tool (even from external providers)?

**Comment by nyalldawson** [comment]
>  sorry for the naive question, does this applies only to qgis native algs or also any other tool (even from external providers)?

Not naive at all! It should apply (transparently) to any providers.

**Comment by nyalldawson** [comment]
@gioman When this is reviewed and merged I'll pull in the commits from https://github.com/nyalldawson/QGIS/tree/more_processing which bring across 2 gdal algs. That should open the door for the rest to be ported :)

**Comment by gioman** [comment]
> Not naive at all! It should apply (transparently) to any providers.

I can't find a good enough emoji to react to this... thanks a thousand tons!

**Comment by nyalldawson** [comment]
> I can't find a good enough emoji to react to this... thanks a thousand tons!

Before you get too excited - keep in mind that this doesn't "multithread" models. So a whole model is now run in a background thread, but there's no logic (yet) to run non-dependant parts in parallel. Unfortunately that part is out of scope for my current work, but should be doable later without API break issues.

**Comment by gioman** [comment]
> Before you get too excited - keep in mind that this doesn't "multithread" models. So a whole model is now run in a background thread, but there's no logic (yet) to run non-dependant parts in parallel. Unfortunately that part is out of scope for my current work, but should be doable later without API break issues.

I understand, thanks. But (unless I got it very wrong) this anyway means that any single tool/algorithm will run in the background, not blocking qgis main window, correct?

**Comment by nyalldawson** [comment]
> I understand, thanks. But (unless I got it very wrong) this anyway means that any single tool/algorithm will run in the background, not blocking qgis main window, correct?

Yes. Well.... when run through the algorithm dialog anyway. There's probably other parts i need to switch away from direct `processing.run(...)` commands. But I'll check that when this PR is merged.

I'd love early tests if you feel like building!

**Comment by gioman** [comment]
> Yes. Well.... when run through the algorithm dialog anyway

this is by itself a *huge* step forward.

**Comment by gioman** [comment]
> I'd love early tests if you feel like building!

no problem, I will.

**Comment by wonder-sk** [comment]
@nyalldawson I have had a look at the commits you highlighted - it all looks great.

I would just suggest to add few more bits to the api docs:
- prepare + post-process - what kind of preparation/post-processing to do (e.g. make feature sources from layers?) - to help alg authors understand what to put into those implementations and what not
- is "prepare" allowed to fail? if it fails, what happens next?
- "post-process" - is it called if alg failed? (from what I understand, no)

For your other comment to allow access to layer stores from a different thread, I probably do not understand why do we need it - isn't it enough to just push such layer store to the thread where it will be needed? I think it is still a bad idea in any thread to do anything with QObject subclasses that belong to another thread (or we risk that bad things happen due to unexpected signal/slot behavior).

**Pull request merged**